micro-rdk/src/common/data_manager.rs

```diff
@@ -610,6 +635,117 @@ where
+                                component_name: collector_key.r_name.clone(),
+                                method_name: collector_key.method.to_string(),
+                                ..Default::default()
+                            }),
+                            sensor_contents: vec![readings],
+                        };
+                        match app_client.upload_data(upload_request).await {
+                            Ok(_) => {
+                                #[cfg(feature = "data-upload-hook-unstable")]
+                                unsafe {
+                                    micro_rdk_data_manager_post_upload_hook();
+                                }
+                            }
+                            Err(err) => {
+                                log::error!("error uploading data, failed to upload {:?} readings on this attempt", collectors_len - idx);
+                                return Err(err);
+                            }
+                        };
+                    }
+                    Err(err) => {
+                        log::error!(
+                            "unable to time correct reading for collector {:?}, {:?}",
+                            collector_key,
+                            err
+                        )
+                    }
+                };
+            }
+        }
+        let _ = send_system_event(SystemEvent::DeepSleep(Some(self.sync_interval)), true).await;
+        Ok(())
+    }
+
+    pub fn from_robot_and_config(
+        robot: &LocalRobot,
+        cfg: &RobotConfig,
+        robot_start_time: Instant,
+    ) -> Result<Self, DataManagerError> {
+        if let Some(cfg) = get_data_service_config(cfg)? {
+            let (collectors, sync_interval, part_id) =
+                get_collectors_and_settings_from_robot_and_config(robot, &cfg)?;
+            match sync_interval {
+                None => Err(DataManagerError::MissingSyncInterval),
+                Some(sync_interval) => {
+                    if collectors.iter().any(|c| {
+                        (c.time_interval().as_secs_f32() - sync_interval.as_secs_f32()).abs()
+                            > 0.001
+                    }) {
+                        Err(DataManagerError::CollectionSyncIntervalUnequal(
+                            sync_interval.as_secs(),
+                        ))
+                    } else {
+                        Ok(Self {
+                            collectors,
+                            sync_interval,
+                            robot_start_time,
+                            part_id,
+                        })
+                    }
+                }
+            }
+        } else {
+            Err(DataManagerError::ConfigError)
+        }
+    }
+}
+
+impl PeriodicAppClientTask for DataCollectAndSyncTask {
+    fn name(&self) -> &str {
+        "DataCollectAndSync"
+    }
+    fn get_default_period(&self) -> Duration {
+        Duration::from_secs(0)
```

Comment:
so if the data sync task fails during collection/sync it will run indefinitely?

Reply:
yes, I have a fix for this in the next PR, but I'll bring it into this one